fish-rust/src/env/env_ffi.rs

```diff
@@ -0,0 +1,368 @@
+        fn is_pathvar(&self) -> bool;
+
+        #[cxx_name = "equals"]
+        fn equals_ffi(&self, rhs: &EnvVar) -> bool;
+
+        #[cxx_name = "as_string"]
+        fn as_string_ffi(&self) -> UniquePtr<CxxWString>;
+
+        #[cxx_name = "as_list"]
+        fn as_list_ffi(&self) -> UniquePtr<wcstring_list_ffi_t>;
+
+        #[cxx_name = "to_list"]
+        fn to_list_ffi(&self, out: Pin<&mut wcstring_list_ffi_t>);
+
+        #[cxx_name = "get_delimiter"]
+        fn get_delimiter_ffi(&self) -> wchar_t;
+
+        #[cxx_name = "get_flags"]
+        fn get_flags_ffi(&self) -> u8;
+
+        #[cxx_name = "clone_box"]
+        fn clone_box_ffi(&self) -> Box<EnvVar>;
+
+        #[cxx_name = "env_var_create"]
+        fn env_var_create_ffi(vals: &wcstring_list_ffi_t, flags: u8) -> Box<EnvVar>;
+
+        #[cxx_name = "env_var_create_from_name"]
+        fn env_var_create_from_name_ffi(
+            name: wcharz_t,
+            values: &wcstring_list_ffi_t,
+        ) -> Box<EnvVar>;
+    }
+    extern "Rust" {
+        type EnvNull;
+        #[cxx_name = "getf"]
+        fn getf_ffi(&self, name: &CxxWString, mode: u16) -> *mut EnvVar;
+        #[cxx_name = "get_names"]
+        fn get_names_ffi(&self, flags: u16, out: Pin<&mut wcstring_list_ffi_t>);
+        #[cxx_name = "env_null_create"]
+        fn env_null_create_ffi() -> Box<EnvNull>;
+    }
+
+    extern "Rust" {
+        type Statuses;
+        #[cxx_name = "get_status"]
+        fn get_status_ffi(&self) -> i32;
+
+        #[cxx_name = "get_pipestatus"]
+        fn get_pipestatus_ffi(&self) -> &Vec<i32>;
+
+        #[cxx_name = "get_kill_signal"]
+        fn get_kill_signal_ffi(&self) -> i32;
+    }
+
+    extern "Rust" {
+        #[cxx_name = "EnvDyn"]
+        type EnvDynFFI;
+        fn getf(&self, name: &CxxWString, mode: u16) -> *mut EnvVar;
+        fn get_names(&self, flags: u16, out: Pin<&mut wcstring_list_ffi_t>);
+    }
+
+    extern "Rust" {
+        #[cxx_name = "EnvStackRef"]
+        type EnvStackRefFFI;
+        fn getf(&self, name: &CxxWString, mode: u16) -> *mut EnvVar;
+        fn get_names(&self, flags: u16, out: Pin<&mut wcstring_list_ffi_t>);
+        fn is_principal(&self) -> bool;
+        fn get_last_statuses(&self) -> Box<Statuses>;
+        fn set_last_statuses(&self, status: i32, kill_signal: i32, pipestatus: &CxxVector<i32>);
+        fn set(
+            &self,
+            name: &CxxWString,
+            flags: u16,
+            vals: &wcstring_list_ffi_t,
+        ) -> EnvStackSetResult;
+        fn remove(&self, name: &CxxWString, flags: u16) -> EnvStackSetResult;
+        fn get_pwd_slash(&self) -> UniquePtr<CxxWString>;
+        fn set_pwd_from_getcwd(&self);
+
+        fn push(&mut self, new_scope: bool);
+        fn pop(&mut self);
+
+        // Returns a Box<OwningNullTerminatedArrayRefFFI>.into_raw() cast to a void*.
+        // This is because we can't use the same C++ bindings to a Rust type from two different bridges.
```

Comment:
we can, for example `ParseErrorListFfi` is used by bridges in `parse_constants.rs`, `parse_tree.rs` and `ast.rs`
It needs this incantation:

```rust
unsafe impl cxx::ExternType for OwningNullTerminatedArrayRefFFI<'_> {
    type Id = cxx::type_id!("OwningNullTerminatedArrayRefFFI");
    type Kind = cxx::kind::Opaque;
}
```

Reply:
nice, TIL. Fixed.